### PR TITLE
Allow deps auditing to run on Linux via HOMEBREW_SIMULATE_MACOS_ON_LINUX

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

`brew audit` currently fails on macOS because `audit_deps` is not run on Linux and all auditing on homebrew-core is done on Linux.

This attempts to prevent that happening again by allowing the audit to run under `HOMEBREW_FORCE_HOMEBREW_ON_LINUX`.

This PR reverts 93362547c99d880e59d0e11f4127f01145ea5b1d which removed the conditional. This PR also restores a867e78f62c807b65ff4f80587d1a512252febf0 for `uses_for_macos` handling, with a fix for it not working due to the `:since` handling. In addition, I've added handling of `on_os` blocks.

To ensure this works, I've added a step to the Linux CI.

See: https://github.com/Homebrew/homebrew-core/pull/70826#issuecomment-789300223

### CODE REVIEW

**Comment by Bo98** [comment]
If I'm not here: `pr-publish`/automerge https://github.com/Homebrew/homebrew-core/pull/72357 to fix the audit that's currently failing.

**Comment by MikeMcQuaid** [comment]
> `brew audit` currently fails on macOS because `audit_deps` is not run on Linux and all auditing on homebrew-core is done on Linux.

To be exact: the tap-wide auditing is done only on macOS but individual `brew audit` of modified formulae is still done on macOS.

Can you remove this auditing from the `test everything (macOS)` job if it's no longer reliable/needed?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -104,9 +104,14 @@ jobs:
           # brew style doesn't like world writable directories
           sudo chmod -R g-w,o-w "$HOMEBREW_REPOSITORY/Library/Taps"
 
+          echo "HOMEBREW_FORCE_HOMEBREW_ON_LINUX=1" >> $GITHUB_ENV
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:107]: Can this be set instead using `env:` through the actions configuration (like other variables)?

**Comment by Bo98** [comment]
> individual `brew audit` of modified formulae is still done on macOS.

Yes, indeed - I made that statement too broad.

`audit_deps` only reliably works on tap-wide auditing, since formulae not directly modified by the PR can be affected. The Qt6 PR is an example or this.

> Can you remove this auditing from the `test everything (macOS)` job if it's no longer reliable/needed?

It is reliable, but the behaviour on Linux should now match macOS so I suppose I can remove it as unneeded.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (Bo98) [reply .github/workflows/tests.yml:107]: I suppose it's only needed for the `brew audit` step - it just seemed more "correct" to have it set for all steps after changing the homebrew-core origin.

**Comment by Bo98** [comment]
> the behaviour on Linux should now match macOS so I suppose I can remove it as unneeded.

Actually, Cask auditing is not included in Linux CI so shouldn't remove this yet.

**Comment by Bo98** [comment]
I think `HOMEBREW_FORCE_HOMEBREW_ON_LINUX` is getting too broad. We're using it a bit to emulate macOS logic on Linux, but we're also using it on Linux as a part of migrating to homebrew-core.

We should probably separate this and introduce a `HOMEBREW_SIMULATE_MACOS_ON_LINUX` or something.

**Comment by MikeMcQuaid** [comment]
> We should probably separate this and introduce a `HOMEBREW_SIMULATE_MACOS_ON_LINUX` or something.

Works for me. 

> I think `HOMEBREW_FORCE_HOMEBREW_ON_LINUX` is getting too broad. We're using it a bit to emulate macOS logic on Linux, but we're also using it on Linux as a part of migrating to homebrew-core.

The original intent was pretty much just to use homebrew-core on Linuxbrew.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:107]: Yeh, just use `env` on those steps (or that step) then I think

**Review by MikeMcQuaid — APPROVED** [review]
Looks good. I wonder if any other `HOMEBREW_FORCE_HOMEBREW_ON_LINUX` uses warrant replacing with the new variable?

**Comment by Bo98** [comment]
I think the other references are fine as far as I can tell.

File: Library/Homebrew/env_config.rb [hunk 2]
```diff
@@ -260,6 +260,11 @@ module EnvConfig
         description: "If set, use Pry for the `brew irb` command.",
         boolean:     true,
       },
+      HOMEBREW_SIMULATE_MACOS_ON_LINUX:       {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:263]: Is there any situation where you'd want just `HOMEBREW_SIMULATE_MACOS_ON_LINUX` and not `HOMEBREW_FORCE_HOMEBREW_ON_LINUX`? If not, could we make `HOMEBREW_SIMULATE_MACOS_ON_LINUX` automatically set/imply `HOMEBREW_FORCE_HOMEBREW_ON_LINUX`?

**Review by MikeMcQuaid — APPROVED** [review]
Looking good! One question, non-blocking, feel free to merge without if desired.

File: Library/Homebrew/env_config.rb (hunk 2 above)
Comment (Bo98) [reply Library/Homebrew/env_config.rb:263]: > Is there any situation where you'd want just `HOMEBREW_SIMULATE_MACOS_ON_LINUX` and not `HOMEBREW_FORCE_HOMEBREW_ON_LINUX`?

Probably not - or at least not for any of our use cases. In theory, `HOMEBREW_SIMULATE_MACOS_ON_LINUX` should work on `linuxbrew-core` and pass `brew audit` without issue - but that's not something we particularly care about.

> make `HOMEBREW_SIMULATE_MACOS_ON_LINUX` automatically set/imply `HOMEBREW_FORCE_HOMEBREW_ON_LINUX`?

Yeah that makes sense.

Comment (MikeMcQuaid) [reply Library/Homebrew/env_config.rb:263]: > Yeah that makes sense.

👍🏻 cool, let's do/document that then.

**Pull request merged**